osf/metrics.py

```diff
@@ -0,0 +1,129 @@
+from elasticsearch_metrics import metrics
+from django.db import models
+
+class MetricMixin(object):
+
+    @classmethod
+    def _get_id_to_count(cls, size, metric_field, count_field, after=None):
+        """Performs the elasticsearch aggregation for get_top_by_count. Return a
+        dict mapping ids to summed counts. If there's no data in the ES index, return None.
+        """
+        search = cls.search()
+        if after:
+            search = search.filter('range', timestamp={'gte': after})
+        search.aggs.\
+            bucket('by_id', 'terms', field=metric_field, order={'sum_count': 'desc'}).\
+            metric('sum_count', 'sum', field=count_field)
+        # Optimization: set size to 0 so that hits aren't returned (we only care about the aggregation)
+        response = search.extra(size=0).execute()
+        # No indexed data
+        if not hasattr(response.aggregations, 'by_id'):
+            return None
+        buckets = response.aggregations.by_id.buckets
+        # Map _id => count
+        return {
+            bucket.key: int(bucket.sum_count.value)
+            for bucket in buckets
+        }
+
+    @classmethod
+    def get_top_by_count(cls, qs, model_field, metric_field,
+                         size=None, order_by=None,
+                         count_field='count',
+                         annotation='metric_count', after=None):
+        """Return a queryset annotated with the metric counts for each item.
+
+        Example: ::
+
+            # Get the top 10 PreprintProviders by download count
+            top_providers = PreprintDownload.get_top_by_count(
+                qs=PreprintProvider.objects.all(),
+                model_field='_id',
+                metric_field='provider_id',
+                annotation='download_count',
+                size=10
+            )
+
+            for each in top_providers:
+                print('{}: {}'.format(each._id, each.download_count))
+
+        ``size`` determines the number of buckets returned by the aggregation.
+        If ``size=None``, the size of the queryset is used.
+        WARNING: Be careful when using size=None when using a large queryset.
```

Comment:
Might make sense to remove the `size=None` default, so it's harder to trip on this accidentally?

Reply:
The API code depends on the `size=None` behavior. This makes it so that the number of items requested from Elasticsearch is equal to the number of items requested from the database.

This shouldn't be a problem, since IIRC we limit the page size in the API to 100, with a few exceptions where it's a bit larger.

Reply:
That makes sense; I just meant remove the default value, so the api code has to explicitly pass `size=None`. Thinking it'll force future code-writers to consider whether they really want a bucket per item, or could do with fewer.